### DISCUSSION TITLE
CompElast = 2 in 5MW_Land_BD_DLL_WTurb.fst

### DISCUSSION THREAD

**Discussion by eletroinf** [OpenFAST/openfast#1590] [General]
I'm starting with the 5 MW wind turbine onshore, with openfast 3.50 and Simulink Sfunc (both .dll and Sfunc downloaded from github).
Interesting thing I've noticed, in the r-test original files, when I try to run I got the error reported bellow. If I change CompElast to 1, the simulation works.
So, what changes if CompElast is 1 or 2? Why am I getting this error?

Other thing I've noticed. In the file "NRELOffshrBsline5MW_Onshore_ElastoDyn_BDoutputs.dat", I've needed to change the following three lines, adding the prefix "../5MW_Baseline/" or I also generates an error when trying to run the simulation:

"../5MW_Baseline/NRELOffshrBsline5MW_Blade.dat"    BldFile(1)  - Name of file containing properties for blade 1 (quoted string)
"../5MW_Baseline/NRELOffshrBsline5MW_Blade.dat"    BldFile(2)  - Name of file containing properties for blade 2 (quoted string)
"../5MW_Baseline/NRELOffshrBsline5MW_Blade.dat"    BldFile(3)  - Name of file containing properties for blade 3 (quoted string) [unused for 2 blades]

I'm using Simulink 2018b.

Regards,
Ricardo Morim

**************************************************************************************************
 OpenFAST

 Copyright (C) 2023 National Renewable Energy Laboratory
 Copyright (C) 2023 Envision Energy USA LTD

 This program is licensed under Apache License Version 2.0 and comes with ABSOLUTELY NO WARRANTY.
 See the "LICENSE" file distributed with this software for details.
 **************************************************************************************************

 OpenFAST-v3.5.0
 Compile Info:
  - Compiler: Intel(R) Fortran Compiler 2021
  - Architecture: 64 bit
  - Precision: single
  - OpenMP: No
  - Date: May 17 2023
  - Time: 14:27:57
 Execution Info:
  - Date: 05/24/2023
  - Time: 19:47:41-0300

 OpenFAST input file heading:
     FAST Certification Test #26: NREL 5.0 MW Baseline Wind Turbine (Onshore)

 Running ElastoDyn.
 Nodal outputs section of ElastoDyn input file not found or improperly formatted.
 Running BeamDyn.
 Nodal outputs section of BeamDyn input file not found or improperly formatted.
 Running BeamDyn.
 Nodal outputs section of BeamDyn input file not found or improperly formatted.
 Running BeamDyn.
 Nodal outputs section of BeamDyn input file not found or improperly formatted.
 Running AeroDyn.
 AD15 Nodal Outputs: Nodal output section of AeroDyn input file not found or improperly formatted.
 Skipping nodal outputs.
 Warning: Turning off Unsteady Aerodynamics because polar has constant data. (node 1, blade 1)
 Warning: Turning off Unsteady Aerodynamics because polar has constant data. (node 2, blade 1)
 Warning: Turning off Unsteady Aerodynamics because polar has constant data. (node 3, blade 1)
 Warning: Turning off Unsteady Aerodynamics because polar has constant data. (node 4, blade 1)
 Warning: Turning off Unsteady Aerodynamics because polar has constant data. (node 1, blade 2)
 Warning: Turning off Unsteady Aerodynamics because polar has constant data. (node 2, blade 2)
 Warning: Turning off Unsteady Aerodynamics because polar has constant data. (node 3, blade 2)
 Warning: Turning off Unsteady Aerodynamics because polar has constant data. (node 4, blade 2)
 Warning: Turning off Unsteady Aerodynamics because polar has constant data. (node 1, blade 3)
 Warning: Turning off Unsteady Aerodynamics because polar has constant data. (node 2, blade 3)
 Warning: Turning off Unsteady Aerodynamics because polar has constant data. (node 3, blade 3)
 Warning: Turning off Unsteady Aerodynamics because polar has constant data. (node 4, blade 3)
 Running InflowWind.

    Reading a 31x31 grid (145 m wide, 17.5 m to 162.5 m above ground) with a characteristic wind
    speed of 9 m/s. This full-field file was generated by TurbSim (v2.00.07a-bjj, 14-Jun-2016) on
    21-May-2023 at 19:43:50.

    Processed 12323 time steps of 20-Hz grid-field data (616.1 seconds).
 Running ServoDyn.

 FAST_InitializeAll:SrvD_Init:ValidatePrimaryData:Yaw angle and rate are not commanded from
 Simulink model.
 ValidatePrimaryData:HSS brake is not commanded from Simulink model.
 ValidatePrimaryData:Airfoil control is not commanded from Simulink model.
 ValidatePrimaryData:Cable control is not commanded from Simulink model.

Warning: Matching "Goto" for "From" 'nrel_5_mw_wt_v1/From' not found 
> In nrel_5_mw_wt_sim_v1 (line 61) 
Warning: Matching "Goto" for "From" 'nrel_5_mw_wt_v1/From5' not found 
> In nrel_5_mw_wt_sim_v1 (line 61) 
Warning: Matching "From" for "Goto" 'nrel_5_mw_wt_v1/Torque Gerador- Conversor Iq/Goto1' not found 
> In nrel_5_mw_wt_sim_v1 (line 61) 
Warning: Output port 1 of 'nrel_5_mw_wt_v1/From' is not connected. 
> In nrel_5_mw_wt_sim_v1 (line 61) 
Warning: Output port 1 of 'nrel_5_mw_wt_v1/From5' is not connected. 
> In nrel_5_mw_wt_sim_v1 (line 61) 
  Time: 0 of 100 seconds.

  OpenFAST completed.

Error using nrel_5_mw_wt_sim_v1 (line 61)
Error reported by S-function 'FAST_SFunc' in 'nrel_5_mw_wt_v1/FAST Nonlinear Wind Turbine/S-Function':
FAST_Solution:FAST_AdvanceStates:B1:BD_GA2:BD_DynamicSolutionGA2:Solution does not converge after the maximum number of iterations
FAST_AdvanceStates:B2:BD_GA2:BD_DynamicSolutionGA2:Solution does not converge after the maximum number of iterations
FAST_AdvanceStates:B3:BD_GA2:BD_DynamicSolutionGA2:Solution does not converge after the maximum number of iterations

**Comment by jjonkman** [comment]
Dear @eletroinf,

OpenFAST primary input `CompElast` determines which structural model is used for the blades--ElastoDyn (`CompElast` = 1) or BeamDyn (2).

But I'm not sure I fully understand what you are doing.

For ElastoDyn, are you using the input files available here: https://github.com/OpenFAST/r-test/tree/main/glue-codes/openfast/5MW_Land_DLL_WTurb?

For BeamDyn, are you using the input files available here: https://github.com/OpenFAST/r-test/tree/main/glue-codes/openfast/5MW_Land_BD_DLL_WTurb?

Best regards,

**Reply by eletroinf** [reply]
Dear jjonkman

So, if I correctly understand, CompElast flag is used to select the model for the blades.
I'm running the file "5MW_Land_BD_DLL_WTurb.fst"

Thank you for the explanation.

Regards,
Ricardo Morim

**Reply by jjonkman** [reply]
Dear @eletroinf,

You should have no problem if you use the r-tests I reference above for ElastoDyn and BeamDyn.  You can file compare the OpenFAST primary (_.fst_) file and ElastoDyn input files to see what their differences are.

Best regards,
